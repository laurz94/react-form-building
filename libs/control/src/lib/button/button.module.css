.lib-button {
  border: none;
  border-radius: var(--border-radius);
  font-variant: small-caps;
  font-size: 1rem;
  padding: 0.5rem;

  &[title] {
    color: fuchsia;
  }

  &.icon-only {
    border-radius: 50%;
    height: 2rem;
    width: 2rem;
  }

  &.primary {
    background-color: var(--primary-color);
    color: var(--text-on-primary-color);

    &:hover {
      background-color: var(--primary-hover-color);
    }

    &:disabled {
      background-color: var(--disabled-control-color);
      color: var(--disabled-control-text-color);
    }
  }

  &.secondary {
    background-color: transparent;
    color: var(--secondary-color);
    border: 1px solid var(--secondary-color);

    &:hover {
      background-color: var(--secondary-hover-color);
    }

    &:disabled {
      background-color: var(--disabled-control-color);
      color: var(--disabled-control-text-color);
      border: none;
    }
  }

  &.tertiary {
    background-color: var(--text-on-tertiary-color);
    color: var(--tertiary-color);

    &:hover {
      background-color: var(--tertiary-hover-color);
    }

    &:disabled {
      background-color: var(--disabled-control-color);
      color: var(--disabled-control-text-color);
    }
  }

  .icon {
    margin-right: 0.25rem;
  }
}
