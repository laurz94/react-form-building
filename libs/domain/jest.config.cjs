const { readFileSync } = require('fs');
const path = require('path');

const swcJestConfig = JSON.parse(
  readFileSync(path.join(__dirname, '.spec.swcrc'), 'utf-8')
);
swcJestConfig.swcrc = false;

module.exports = {
  displayName: 'domain',
  preset: '../../jest.preset.js',
  testEnvironment: 'node',
  transform: {
    '^.+\\.[tj]s$': ['@swc/jest', swcJestConfig],
  },
  moduleFileExtensions: ['ts', 'js', 'html'],
  coverageDirectory: 'test-output/jest/coverage',
};
