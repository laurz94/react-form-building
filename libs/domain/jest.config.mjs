import { readFileSync } from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';

// Derive __dirname from import.meta.url
const __dirname = path.dirname(fileURLToPath(import.meta.url));

// Reading the SWC compilation config for the spec files
const swcJestConfig = JSON.parse(
  readFileSync(`${__dirname}/.spec.swcrc`, 'utf-8')
);

// Disable .swcrc look-up by SWC core because we're passing in swcJestConfig ourselves
swcJestConfig.swcrc = false;

export default {
  displayName: 'domain',
  preset: '../../jest.preset.js',
  testEnvironment: 'node',
  transform: {
    '^.+\\.[tj]s$': ['@swc/jest', swcJestConfig],
  },
  moduleFileExtensions: ['ts', 'js', 'html'],
  coverageDirectory: 'test-output/jest/coverage',
};
