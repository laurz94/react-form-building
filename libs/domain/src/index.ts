export * from './lib/constants/index;
export * from './lib/functions/index';
export * from './lib/models/index';
