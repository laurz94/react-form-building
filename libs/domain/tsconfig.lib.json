{
  "extends": "../../tsconfig.base.json",
  "compilerOptions": {
    "baseUrl": ".",
    "rootDir": "src",
    "outDir": "dist",
    "tsBuildInfoFile": "dist/tsconfig.lib.tsbuildinfo",
    "emitDeclarationOnly": true,
    "forceConsistentCasingInFileNames": true,
    "types": ["node"]
  },
  "include": ["src/**/*.ts", "../control/src/lib/date-picker/date-picker.configuration.ts", "../control/src/lib/number/number.configuration.ts", "../control/src/lib/textbox/textbox.configuration.ts"],
  "references": [],
  "exclude": ["jest.config.ts", "src/**/*.spec.ts"]
}
