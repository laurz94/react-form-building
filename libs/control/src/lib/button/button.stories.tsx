import type { Meta, StoryObj } from '@storybook/react';
import { LibButton } from './button';
import { within } from '@storybook/testing-library';
import { expect } from '@storybook/jest';

const meta: Meta<typeof LibButton> = {
  component: LibButton,
  title: 'LibButton',
};
export default meta;
type Story = StoryObj<typeof LibButton>;

export const Primary = {
  args: {},
};

export const Heading: Story = {
  args: {},
  play: async ({ canvasElement }) => {
    const canvas = within(canvasElement);
    expect(canvas.getByText(/Welcome to LibButton!/gi)).toBeTruthy();
  },
};
