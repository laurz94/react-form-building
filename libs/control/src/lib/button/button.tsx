import { ButtonConfiguration } from './button.configuration';
import styles from './button.module.css';

export function LibButton({
  icon,
  id,
  isDisabled,
  isLoading,
  label,
  type,
  onClicked,
}: ButtonConfiguration) {
  function getIcon() {
    if (isLoading) {
      return (
        <i
          className={'icon fa-solid fa-spinner fa-spin-pulse' + styles['icon']}
          data-testid={id + '-button-loading-icon'}
        ></i>
      );
    } else if (icon) {
      return (
        <i
          className={icon + ' ' + styles['icon']}
          data-testid={id + '-button-icon'}
        ></i>
      );
    } else {
      return '';
    }
  }

  return (
    <button
      type="button"
      className={`${styles['lib-button']} ${styles[type]}`}
      id={id + '-button'}
      data-testid={id + '-button'}
      disabled={isDisabled}
      onClick={onClicked}
    >
      {getIcon()}
      {label}
    </button>
  );
}
