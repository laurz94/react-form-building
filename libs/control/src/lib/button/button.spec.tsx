/**
 * @jest-environment jsdom
 */

import { ButtonConfiguration } from '@libs/domain';
import '@testing-library/jest-dom';
import { act } from 'react';
import ReactDOMClient from 'react-dom/client';
import { LibButton } from './button';

describe('Button', () => {
  async function getButton(config: ButtonConfiguration) {
    const container = document.createElement('div');
    document.body.appendChild(container);

    await act(async () =>
      ReactDOMClient.createRoot(container).render(
        <LibButton
          id={config.id}
          isDisabled={config.isDisabled}
          isLoading={config.isLoading}
          label={config.label}
          icon={config.icon}
          type={config.type}
        />
      )
    );

    return container.querySelector('#test-button');
  }
  function expectToMatchSnapshot(button: Element | null) {
    expect(button).toBeTruthy();
    expect(button).toMatchSnapshot();
  }

  describe('primary button', () => {
    it('can render primary button with only label', async () => {
      const button = await getButton({ type: 'primary', id: 'test', label: 'test' });
      expectToMatchSnapshot(button);
    });
    it('can render primary button with only icon', async () => {
      const button = await getButton({
        type: 'primary',
        id: 'test',
        icon: 'fa-solid fa-floppy-disk',
      });
      expectToMatchSnapshot(button);
    });
    it('can render primary button with label and icon', async () => {
      const button = await getButton({
        type: 'primary',
        id: 'test',
        icon: 'fa-solid fa-floppy-disk',
        label: 'test',
      });
      expectToMatchSnapshot(button);
    });
  });

  describe('secondary button', () => {
    it('can render secondary button with only label', async () => {
      const button = await getButton({ type: 'secondary', id: 'test', label: 'test' });
      expectToMatchSnapshot(button);
    });
    it('can render secondary button with only icon', async () => {
      const button = await getButton({
        type: 'secondary',
        id: 'test',
        icon: 'fa-solid fa-floppy-disk',
      });
      expectToMatchSnapshot(button);
    });
    it('can render secondary button with label and icon', async () => {
      const button = await getButton({
        type: 'secondary',
        id: 'test',
        icon: 'fa-solid fa-floppy-disk',
        label: 'test',
      });
      expectToMatchSnapshot(button);
    });
  });

  describe('tertiary button', () => {
    it('can render tertiary button with only label', async () => {
      const button = await getButton({ type: 'tertiary', id: 'test', label: 'test' });
      expectToMatchSnapshot(button);
    });
    it('can render tertiary button with only icon', async () => {
      const button = await getButton({
        type: 'tertiary',
        id: 'test',
        icon: 'fa-solid fa-floppy-disk',
      });
      expectToMatchSnapshot(button);
    });
    it('can render tertiary button with label and icon', async () => {
      const button = await getButton({
        type: 'tertiary',
        id: 'test',
        icon: 'fa-solid fa-floppy-disk',
        label: 'test',
      });
      expectToMatchSnapshot(button);
    });
  });

  describe('loading icon', () => {
    it('show the loading icon instead of the icon', async() => {
      const button = await getButton({
        type: 'primary',
        id: 'test',
        isLoading: true,
        icon: 'fa-solid fa-floppy-disk',
        label: 'test',
      });
      const loadingIcon = button?.querySelector('[data-testid="test-button-loading-icon"]');
      const icon = button?.querySelector('[data-testid="test-button-icon"]');

      expect(loadingIcon).toBeTruthy();
      expect(icon).toBeFalsy();
    });
  });

  describe('is disabled', () => {
    it('be disabled when set', async() => {
      const button = await getButton({
        type: 'primary',
        id: 'test',
        isDisabled: true,
        icon: 'fa-solid fa-floppy-disk',
        label: 'test',
      });
      
      expect(button).toBeDisabled();
    });
    it('be enabled when set', async() => {
      const button = await getButton({
        type: 'primary',
        id: 'test',
        isDisabled: false,
        icon: 'fa-solid fa-floppy-disk',
        label: 'test',
      });
      
      expect(button).toBeEnabled();
    });
  });
});
