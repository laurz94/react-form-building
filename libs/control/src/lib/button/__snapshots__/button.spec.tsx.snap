// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button primary button can render primary button with label and icon 1`] = `
<button
  class="lib-button primary"
  data-testid="test-button"
  id="test-button"
  type="button"
>
  <i
    class="fa-solid fa-floppy-disk icon"
    data-testid="test-button-icon"
  />
  test
</button>
`;

exports[`Button primary button can render primary button with only icon 1`] = `
<button
  class="lib-button primary"
  data-testid="test-button"
  id="test-button"
  type="button"
>
  <i
    class="fa-solid fa-floppy-disk icon"
    data-testid="test-button-icon"
  />
</button>
`;

exports[`Button primary button can render primary button with only label 1`] = `
<button
  class="lib-button primary"
  data-testid="test-button"
  id="test-button"
  type="button"
>
  test
</button>
`;

exports[`Button secondary button can render secondary button with label and icon 1`] = `
<button
  class="lib-button secondary"
  data-testid="test-button"
  id="test-button"
  type="button"
>
  <i
    class="fa-solid fa-floppy-disk icon"
    data-testid="test-button-icon"
  />
  test
</button>
`;

exports[`Button secondary button can render secondary button with only icon 1`] = `
<button
  class="lib-button secondary"
  data-testid="test-button"
  id="test-button"
  type="button"
>
  <i
    class="fa-solid fa-floppy-disk icon"
    data-testid="test-button-icon"
  />
</button>
`;

exports[`Button secondary button can render secondary button with only label 1`] = `
<button
  class="lib-button secondary"
  data-testid="test-button"
  id="test-button"
  type="button"
>
  test
</button>
`;

exports[`Button tertiary button can render tertiary button with label and icon 1`] = `
<button
  class="lib-button tertiary"
  data-testid="test-button"
  id="test-button"
  type="button"
>
  <i
    class="fa-solid fa-floppy-disk icon"
    data-testid="test-button-icon"
  />
  test
</button>
`;

exports[`Button tertiary button can render tertiary button with only icon 1`] = `
<button
  class="lib-button tertiary"
  data-testid="test-button"
  id="test-button"
  type="button"
>
  <i
    class="fa-solid fa-floppy-disk icon"
    data-testid="test-button-icon"
  />
</button>
`;

exports[`Button tertiary button can render tertiary button with only label 1`] = `
<button
  class="lib-button tertiary"
  data-testid="test-button"
  id="test-button"
  type="button"
>
  test
</button>
`;
